survey/models/survey.py

```diff
@@ -1,9 +1,66 @@
 # -*- coding: utf-8 -*-
+from datetime import datetime
 
 from django.db import models
 from django.urls import reverse
 from django.utils.translation import gettext_lazy as _
 
+class ActiveManager(models.Manager):
+    """
+    This special manager only retrieves active objects (when the current date
+    is between the object's publish date and/or its expiration date).
+
+    The date fields are given when creating the Manager instance. If either is
+    None then the manager will not take it into account for filtering.
+
+    Example definition for a model:
+
+    class ExampleModel(models.Model):
+        publish_date = models.DateTimeField()
+        expire_date = models.DateTimeField(blank=True, null=True)
+
+        actives = ActiveManager(from_date='publish_date', to_date='expire_date')
+
+    Or if the model only had an expire_date:
+
+    class ExampleModel(models.Model):
+        expire_date = models.DateTimeField(blank=True, null=True)
+
+        actives = ActiveManager(from_date=None, to_date='expire_date')
+
+    Each instance of the manager has an attribute date_filters which can be used in
+    custom queries. For example, if you have a ManyToMany relationship to a model
+    with ActiveManager and you can't access via the manager (because you
+    need to use select_related, for example) then you can do:
+
+    instance.many_to_many.filter(*ExampleModel.actives.date_filters)
+
+    """
+
+    def __init__(self, from_date=None, to_date=None):
+        super(ActiveManager, self).__init__()
+        self.from_date = from_date
+        self.to_date = to_date
+        now = datetime.now
+        if from_date and to_date:
+            self.date_filters = (models.Q(**{'%s__isnull' % self.to_date: True}) |
+                                 models.Q(**{'%s__gte' % self.to_date: now}),
+                                 models.Q(**{'%s__isnull' % self.from_date: True}) |
+                                 models.Q(**{'%s__lte' % self.from_date: now}))
+
+        elif from_date:
+            self.date_filters = (models.Q(**{'%s__isnull' % self.from_date: True}) |
+                                 models.Q(**{'%s__lte' % self.from_date: now}),)
+        elif to_date:
+            self.date_filters = (models.Q(**{'%s__isnull' % self.to_date: True}) |
+                                 models.Q(**{'%s__gte' % self.to_date: now}),)
+        else:
+            raise ValueError # "At least one date field is required"
```

Comment:
yes, but there is always the creation date (publish_date) which is set transparently in the background. In an upcoming enhancement you could make a survey be only visible in the future, by changing the auto_now argument. I did not do it because I like it to be set to today.

Reply:
If there would be no date, it would fail the loading of the site. Also it is failing for old surveys, because there are no datefields in the database for them! <- this will be an issue for updating running surveys, I think...
If dates are set to none it should be equivalent to always.

Reply:
What if we put the default value to "None" and it means there is no time restriction for the past? It would effectively have the same effect than putting `datetime.now()` but without calling it, and with easier migration of old surveys.

Reply:
Maybe we should put this patch in hold. I'll work on this some more and will try to unify expiry and is_publisher!

Reply:
Sounds good to me :) Don't worry there is no rush, thanks again for handling this.